kernel/kernel-api/src/main/java/io/delta/kernel/types/StructType.java

```diff
@@ -40,6 +41,22 @@ public StructType() {
   }
 
   public StructType(List<StructField> fields) {
+    validateNoNestedMetadataColumns(fields);
```

Comment:
This method will repeatedly get called for every `add` call, but I don't see a way to prevent this as long as the constructor is public.